### PR TITLE
Multiple variable initializers in for loop

### LINKED ISSUE #83: Multiple variable initializers in for loop (`for (var x, y = 0; x; x++) { .. }`)

### PR DESCRIPTION
**Pull request by chalin**
Fixes #83.

### CODE REVIEW

File: test/DartTest.ts [hunk 1]
```diff
@@ -41,8 +41,19 @@ describe('transpile to dart', () => {
       expectTranslate('var a;').to.equal(' var a ;');
       expectTranslate('var a:any;').to.equal(' dynamic a ;');
     });
-    it('should transpile variable declaration lists',
-       () => { expectTranslate('var a: number, b: string;').to.equal(' num a ; String b ;'); });
+    it('should transpile variable declaration lists', () => {
+      expectTranslate('var a, b;').to.equal(' var a , b ;');
+      expectTranslate('var a: number, b;').to.equal(' num a , b ;');
```
Comment (mprobst) [inline test/DartTest.ts]: doesn't this mean that 'b' is also a num in Dart?

File: main.ts [hunk 2]
```diff
@@ -217,13 +217,17 @@ class Translator {
...
+        // Note that a VariableDeclarationList can only occur as part of a for loop.
         var varDeclList = <ts.VariableDeclarationList>node;
-        this.visitList(varDeclList.declarations, ';');
+        this.visitList(varDeclList.declarations, ',');
         break;
 
       case ts.SyntaxKind.VariableDeclaration:
         var varDecl = <ts.VariableDeclaration>node;
-        if (varDecl.type) {
+        if (varDecl.parent.declarations[0] !== varDecl) {
```
Comment (mprobst) [inline main.ts]: maybe it'd be more readable to put this check into the `if (varDecl.type)` branch below?

File: test/DartTest.ts (hunk 1 above)
Comment (chalin) [inline test/DartTest.ts]: Yes, as in other languages (C, Java, C#). I thought that that it is what it meant in TS, but apparently not. Very good catch!

**Comment by chalin** [comment]
Here is a commit which implements the assumptions presented in https://github.com/angular/ts2dart/issues/83#issuecomment-85282233.

File: main.ts [hunk 3]
```diff
@@ -217,16 +217,32 @@ class Translator {
         break;
 
       case ts.SyntaxKind.VariableDeclarationList:
+        // Note that a VariableDeclarationList can only occur as part of a for loop.
         var varDeclList = <ts.VariableDeclarationList>node;
-        this.visitList(varDeclList.declarations, ';');
+        this.visitList(varDeclList.declarations, ',');
```
Comment (mprobst) [inline main.ts]: the comma is the default separator in list, so you can just not specify the second param.

File: main.ts [hunk 4]
```diff
@@ -217,16 +217,32 @@ class Translator {
...
+         * All variables in a variable declaration list must have the same type. Since
+         * we are doing syntax directed translation, we cannot reliably determine if distinct
+         * variables are declared with the same type or not. Hence we support the following cases:
+         * - A variable declaration list with a single variable can be explicitly typed.
+         * - When more than one variable is in the list, all must be implicitly typed.
+         */
+        var firstDecl = varDecl.parent.declarations[0];
+        var msg = 'Variables in a declaration list of more than one variable cannot by typed';
+        if (firstDecl === varDecl) {
+          if (!varDecl.type) {
```
Comment (mprobst) [inline main.ts]: Could you extract this to a method "visitVarDeclType"?

Then you could also use early exits from the function to make the boolean conditions a bit easier to read.

File: test/DartTest.ts [hunk 5]
```diff
@@ -41,8 +41,21 @@ describe('transpile to dart', () => {
...
+    it('should transpile variable declaration lists with initializers', () => {
+      expectTranslate('var a = 0;').to.equal(' var a = 0 ;');
+      expectTranslate('var a, b = 0;').to.equal(' var a , b = 0 ;');
+      expectTranslate('var a = 1, b = 0;').to.equal(' var a = 1 , b = 0 ;');
+    });
+    it('does not support declaration lists containing more than one type (implicit or explicit)', () => {
+      var msg = 'Variables in a declaration list of more than one variable cannot by typed';
+      expectErroneousCode('var a: A, b;').to.throw(msg);
+      expectErroneousCode('var a, b: B;').to.throw(msg);
+      expectErroneousCode('var a: number, b: string;').to.throw(msg);
```
Comment (mprobst) [inline test/DartTest.ts]: Nice tests. I'd add one more case:
```var untyped, a: number, b: string;```

File: main.ts (hunk 4 above)
Comment (chalin) [inline main.ts]: Good suggestion.

**Comment by chalin** [comment]
Changes done. Note that the new visit method has signature `visitVariableDeclarationType(varDecl: ts.VariableDeclaration)` rather than `visitVariableDeclarationType(t: ts.TypeNode)` because a parent cannot be extracted from `t` (to get the `varDecl`) when it is absent.

**Comment by mprobst** [comment]
I've merged this in 27eaa56.

**Pull request closed without merging**